src/ray/raylet/local_task_manager.cc

```diff
@@ -765,16 +747,18 @@ void ReplyCancelled(std::shared_ptr<internal::Work> &work,
 }
 }  // namespace
 
-bool LocalTaskManager::CancelTask(
-    const TaskID &task_id,
+bool LocalTaskManager::CancelTasks(
```

Comment:
I have seen this idiom of removing a task in a complex data structure, by playing with iterators and `erase` in Ray in many places. I feel we can try to split "cancel the task" and "remove it from data structure" by using `std::erase_if`. However it's in C++20 so we can make a poly fill. Here is what I get:

In container_util.h:
```cc

/// Polyfill for C++20 std::erase_if.
/// It's regrettable it's not landed in C++17 or earlier.
template <typename C, typename Pred>
size_t erase_if_polyfill(C &c, Pred pred) {
  auto it = std::remove_if(c.begin(), c.end(), pred);
  size_t count = std::distance(it, c.end());
  c.erase(it, c.end());
  return count;
}

```

Here:
```cc
#include "ray/util/container_util.h"

bool LocalTaskManager::CancelTasks(
    std::function<bool(const RayTask &)> predicate,
    rpc::RequestWorkerLeaseReply::SchedulingFailureType failure_type,
    const std::string &scheduling_failure_message) {
  size_t cancelled_from_to_dispatch = 0;
  size_t sched_cls_removed = erase_if_polyfill(
      tasks_to_dispatch_,
      [&predicate,
       &cancelled_from_to_dispatch,
       failure_type,
       scheduling_failure_message,
       this](auto &shape) -> bool {
        auto &[sched_cls, work_queue] = shape;
        size_t work_removed = erase_if_polyfill(
            work_queue,
            [&predicate, failure_type, scheduling_failure_message, this](
                std::shared_ptr<internal::Work> &work) -> bool {
              const auto &task = work->task;
              const bool should_cancel = predicate(task);
              if (should_cancel) {
                const TaskID task_id = task.GetTaskSpecification().TaskId();
                RAY_LOG(DEBUG) << "Canceling task " << task_id << " from dispatch queue.";
                ReplyCancelled(work, failure_type, scheduling_failure_message);
                if (work->GetState() == internal::WorkStatus::WAITING_FOR_WORKER) {
                  // We've already acquired resources so we need to release them.
                  cluster_resource_scheduler_->GetLocalResourceManager()
                      .ReleaseWorkerResources(work->allocated_instances);
                  // Release pinned task args.
                  ReleaseTaskArgs(task_id);
                }
                if (!task.GetTaskSpecification().GetDependencies().empty()) {
                  task_dependency_manager_.RemoveTaskDependencies(
                      task.GetTaskSpecification().TaskId());
                }
                RemoveFromRunningTasksIfExists(task);
                work->SetStateCancelled();
              }
              // Erase from work_queue if the work is cancelled.
              return should_cancel;
            });
        cancelled_from_to_dispatch += work_removed;
        // Erase from tasks_to_dispatch_ if the work_queue is empty.
        return work_queue.empty();
      });

  size_t cancelled_from_waiting_task_queue = erase_if_polyfill(
      waiting_task_queue_,
      [&predicate, failure_type, scheduling_failure_message, this](
          std::shared_ptr<internal::Work> &work) -> bool {
        const auto &task = work->task;
        const bool should_cancel = predicate(task);
        if (should_cancel) {
          ReplyCancelled(work, failure_type, scheduling_failure_message);
          if (!task.GetTaskSpecification().GetDependencies().empty()) {
            task_dependency_manager_.RemoveTaskDependencies(
                task.GetTaskSpecification().TaskId());
          }
          waiting_tasks_index_.erase(task.GetTaskSpecification().TaskId());
        }
        return should_cancel;
      });
  // Maybe we can return a size_t of num of work cancelled in CancelTasks.
  return (cancelled_from_to_dispatch + cancelled_from_waiting_task_queue > 0);
}

```

Reply:
I don't think std::remove_if works with hash_map: https://stackoverflow.com/questions/9210014/remove-elements-from-an-unordered-map-fulfilling-a-predicate